FileService_AddNewFile

mekk1t opened the pull request:
Implemented a file addition with its the only yet test called "SuccessfulAddNewFileTest".

mekk1t committed:
Created Command

Created a command named "AddNewFileCommand" with its interface.
It validates the incoming request then maps it to the Database model. That model is passed into the repository AddNewFile method.

mekk1t committed:
Create FileController.cs

mekk1t committed:
Create Mapper

Created a mapper interface with a generic method declaration
The mapper itself implements both mapping from DbFile to FileRequest and from FileRequest to DbFile, depending on the input data.

mekk1t committed:
Created Repository

Created a repo and its interface.

mekk1t committed:
Create File Validator

Created a file validator.

I have some doubts about how the .Must() is working, haven't tested it yet.
But later I'll add Validator tests and see whether it's working.

mekk1t committed:
Classes Configuration

Configured Commands, Mappers, Repositories and Validators in Startup.cs

Moved private methods to the top of the class.

mekk1t committed:
FluentValidation Installation + New Model

Installed a NuGet package FluentValidation and created a new model FileRequest that we get from the client / GUI

mekk1t committed:
UnitTests Setup

Installed all required NuGet packages: FluentValidation, EFCore.InMemory, Moq for the future tests.

Created empty classes in the respective folders.

mekk1t committed:
Test 1: Successful File Addition

Created my first UnitTest and it seems to pass.
Debugging also shows that the output Id matches the request's Id.

mekk1t commented:
Other tests will be added later in the evening.

Meanwhile, please, check the only test if it's correct.
Well, it's passing but is it correctly designed?

DarkFelas commented on DigitalOffice/Services/FileService/Models/FileRequest.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        public Guid Id { get; set; }
+        public string Content { get; set; }
+        public string ContentExtension { get; set; }
```
Rename -> Extension

mekk1t replied:
Fixed, migrated

DarkFelas commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -1,41 +1,80 @@
         }
-        
+
+        private void ConfugureTransientCommands(IServiceCollection services)
```
Rename -> ConfigureCommands
Change other methods as well

mekk1t replied:
Renamed, my mistake

DarkFelas commented on DigitalOffice/Services/FileService/Mappers/FileMapper.cs:
```diff
@@ -0,0 +1,32 @@
+        {
+            return new DbFile()
+            {
+                Id = file.Id,
```
Add new property to DbFile -> DateTime Added = DateTime.UtcNow

mekk1t replied:
Should we add this to FileRequest as well? Does front-end need to know when the file was added to Db?
Something like "you uploaded this on 01.01.01", or "uploaded on 31.12.2000"

KostinP replied:
> Should we add this to FileRequest as well? Does front-end need to know when the file was added to Db?
> Something like "you uploaded this on 01.01.01", or "uploaded on 31.12.2000"

Do you think about the situation "when the backend lags"?

mekk1t replied:
> > Should we add this to FileRequest as well? Does front-end need to know when the file was added to Db?
> > Something like "you uploaded this on 01.01.01", or "uploaded on 31.12.2000"
> 
> Do you think about the situation "when the backend lags"?

Didnt quite get your point. What does that situation mean?
You mean that the Date.Now should come from frontend and if its not there then we program the time by ourselves?

KostinP replied:
Not. I thought that you were thinking, what if after sending the file it does not reach the service for a long time. Should we in this case indicate the "time of sending to add" or "time of adding".

And I tried to find out what you meant.

Upd. After your clarification, I understood what you meant. 
If you do not understand what I wrote about, you can write to me in Slack.

DarkFelas replied:
No need, just in dbModel at this moment and use DateTime.**UtcNow**

mekk1t committed:
Removed "Tests" from the folder and namespace names

mekk1t committed:
DbFile 2.0

Updated DbFileby renaming ContentExtension to Extension and adding a new property "AddedOn"
Created a migration and added mapping

mekk1t committed:
Update Startup.cs

Renamed Configure-methods.

mekk1t committed:
Mapper Tests

I did not use SerializerAssert.AreEqual because it is impossible to match file time addition.
It always is created at least 0.000001 ms later or something. And we do not have "AddedOn" property in FileRequest model.

1. NullReference Test: mapper cannot map nothing
2. Request -> Db mapping Test
3. Db -> Request mapping Test

mekk1t committed:
New Command Tests

Add new FailedCommandTests:
1. Duplicate Guid
2. Null Request
3. Invalid file encoding

mekk1t committed:
Validator Tests

+ Renamed validator folder

mekk1t committed:
New Repository Tests

denis-kriv commented on DigitalOffice/Services/FileService/Commands/AddNewFileCommand.cs:
```diff
@@ -0,0 +1,35 @@
+        public Guid Execute(FileRequest request)
+        {
+            validator.ValidateAndThrow(request);
+            DbFile newFile = mapper.Map(request);
```
why not to use `var` here?

mekk1t replied:
why not to use DbFile here?
I don't think its clear what type mapper maps request to. So I clarified it.

denis-kriv replied:
If we will have different mappers, then type of object will be obviously

mekk1t replied:
I renamed request model names so now it's easier to understand what mapper maps to.

denis-kriv commented on DigitalOffice/Services/FileService/Mappers/FileMapper.cs:
```diff
@@ -0,0 +1,33 @@
+using FileService.Models;
+using System;
+
+namespace FileService.Mappers
+{
+    public class FileMapper : IMapper<DbFile, FileRequest>, IMapper<FileRequest, DbFile>
```
maybe we should do two different mappers here (Solid)

mekk1t replied:
Agree. Will fix

mekk1t replied:
@DenisKrivoruchko11 @DarkFelas Spartak advised that it can be implemented via a single mapper.

Come on, think about it. We don't create a whole new repository just for adding files and then another one for editing files.

Solid is actually in play here: S -- Mapper Maps, I - Inherites needed interfaces.

The extra classes are junk and unnecessary I think.

denis-kriv replied:
but we have two methods with name `Map` and I think, it is not good

mekk1t replied:
Yeah, I mentioned above that it's SOLID -- actually, 50/50.
So either way SOLID is broken this time, no matter what we do -- clutter the interfaces or follow Single Responsibility and create extra files.

Inzilbeth replied:
I think it's ok

denis-kriv commented on DigitalOffice/Services/FileService/Models/FileRequest.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace FileService.Models
+{
+    public class FileRequest
```
Maybe `FileCreateRequest` ?

mekk1t replied:
Nah. All Request models are kinda the same, for now at least.

Could you name a reason to have different model names?

denis-kriv replied:
If we want to get a file, it would be logical to call the request getFileRequest if we want to createFileRequest

mekk1t replied:
Alright I'll rename the model,
fixed

denis-kriv commented on DigitalOffice/Tests/FileServiceTests/UnitTestLibrary/SerializerAssert.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace FileServiceUnitTests.UnitTestLibrary
+{
+    public class SerializerAssert
+    {
+        
+    }
```
I can not understand why we need this

mekk1t replied:
Hmm, I deleted this file and it should not be here

MihailKulikov replied:
File still here.

mekk1t replied:
@djino97 @DarkFelas after some thought and conversation with Evgeniy I'm on their side on this one.
I can move a whole lot of code to the static class that will just take two arguments rather than write 5 or how-many-properties rows of code in whichever tests will be added to the FileService, for example.
My only problem with SerializerAssert was that I could not compare two Json objects because of the AddedOn property: each time it was different and never the same in both objects. 

Serializing objects to Json and then comparing is cool -- a few lines of code and we get a quick comparison.

denis-kriv commented on DigitalOffice/Tests/FileServiceTests/Repositories/AddNewFileRepositoryTests.cs:
```diff
@@ -0,0 +1,53 @@
+            };
+        }
+
+        [Test]
+        public void SuccessfulAddNewProjectInDbTest()
+        {
+            Assert.AreEqual(newFile.Id, repository.AddNewFile(newFile));
+        }
```
One test?

mekk1t replied:
Well, what do you suggest to add here?
Im thinking only about Fail tests

mekk1t replied:
But even then all exceptions, checks are passed before dbContext.Add method so i did not include Fail tests.

Could you please name one test that checks something that throws an exception AND that exception would not be thrown by a validator or a mapper that are called before?

mekk1t replied:
Hmm. I've come up with some tests concerning Database exceptions. Will be added some time later.

mekk1t replied:
Added only one test: Failed Add when that file already exists (duplicate Guid); 
other tests like invalid model or model data I think are unnecessary because these errors will be caught by validators or mappers.

I could add Fail Tests in case someone thinks that we can use repository without addressing the data to Validators and Mappers first.

djino97 commented on DigitalOffice/Services/FileService/Commands/AddNewFileCommand.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public Guid Execute(FileRequest request)
+        {
+            validator.ValidateAndThrow(request);
+            DbFile newFile = mapper.Map(request);
```
add one line

mekk1t replied:
fixed

djino97 commented on DigitalOffice/Services/FileService/Repositories/FileRepository.cs:
```diff
@@ -0,0 +1,25 @@
+        public Guid AddNewFile(DbFile file)
+        {
+            dbContext.Files.Add(file);
+            dbContext.SaveChanges();
+            return file.Id;
```
add one line

mekk1t replied:
fixed

djino97 commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -1,41 +1,80 @@
         }
-        
+
+        private void ConfugureCommands(IServiceCollection services)
+        {
+            services.AddTransient<IAddNewFileCommand, AddNewFileCommand>();
+        }
+
+        private void ConfugureRepositories(IServiceCollection services)
+        {
+            services.AddTransient<IFileRepository, FileRepository>();
+        }
+
+        private void ConfugureMappers(IServiceCollection services)
+        {
+            services.AddTransient<IMapper<DbFile, FileRequest>, FileMapper>();
+            services.AddTransient<IMapper<FileRequest, DbFile>, FileMapper>();
+        }
+
+        private void ConfugureValidators(IServiceCollection services)
+        {
+            services.AddTransient<IValidator<FileRequest>, NewFileValidator>();
+        }
```
Move all methods bottom the method ConfigureServices (convention)

mekk1t replied:
Convention? We decided to declare private methods first, didnt we?

Well, I remember declaring private fields first, but this also applies to the methods?

Inzilbeth replied:
"Методы, связанные друг с другом, размещаются по нисходящей: сначала главный метод, потом вызываемый метод, потом другой вызываемый метод и т.д." source: DigitalOffice styleguide

djino97 replied:
@Inzilbeth Yes, so it was written there.

djino97 commented on DigitalOffice/Tests/FileServiceTests/Mappers/FileMapperTests.cs:
```diff
@@ -0,0 +1,91 @@
+
+            Assert.IsInstanceOf<DateTime>(newFile.AddedOn);
+            Assert.IsInstanceOf<Guid>(newFile.Id);
+            Assert.AreEqual(expectedFile.Id, newFile.Id);
+            Assert.AreEqual(expectedFile.Content, newFile.Content);
+            Assert.AreEqual(expectedFile.Extension, newFile.Extension);
+            Assert.AreEqual(expectedFile.IsActive, newFile.IsActive);
```
It is checks can be place in a separate method of class SerializerAssert ).

mekk1t replied:
> It is checks can be place in a separate method of class SerializerAssert ).

I specifically mentioned it in the commit description.
I cannot test it via serializerassert because of the AddedOn property.

djino97 replied:
You can do this only for AreEqual.

mekk1t replied:
So you suggest I add a separate method for these checks only? In a static class? I don't think it's a good idea.

I don't mind doing it though, but I disagree nonetheless.

mekk1t replied:
Fixed it in SerializerAssert.cs

MihailKulikov commented on DigitalOffice/Services/FileService/Mappers/FileMapper.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    public class FileMapper : IMapper<DbFile, FileRequest>, IMapper<FileRequest, DbFile>
+    {
+        public FileRequest Map(DbFile dbFile)
```
For what purpose we want map DbFile to FileRequest?

mekk1t replied:
Well, for the time being we do not have FileResponse model so I treat FileRequest as the only frontend-model.
Also, it is needed when you want to get an actual file. How else are we gonna get it if not map it from DbFile to just File?

Remember, Spartak taught as that we always have different models and transfer data from one thing to another.

MihailKulikov replied:
I think it is wrong to consider FileRequest as the only frontend-model. The model for the response may differ from the model for the request. It would be more correct to create a separate model for the responce, but then this method will be useless.

mekk1t replied:
I agree that it is wrong. I mean that while we do not have Responses, I decided to treat this model as the only one.

Seems I better have created a Response instead

mekk1t replied:
Fixed it, creating a separate model called "File".

Changed mapping from DbFile -> FileRequest to
DbFile -> File

MihailKulikov commented on DigitalOffice/Services/FileService/Mappers/FileMapper.cs:
```diff
@@ -0,0 +1,33 @@
+                Name = dbFile.Name
+            };
+        }
+        public DbFile Map(FileRequest file)
```
I think, we should add here check for null

mekk1t replied:
What kind of check? To throw an exception?
It will already do that for us.
We should add null check when we have an ExceptionHandler in our Middleware. And even after that an exception will be thrown and be caught in the middleware, so checks on the mapper level are unnecessary.

MihailKulikov replied:
As far as I know, it is useful to provide the most specific information with an exception throwing. For example, in this case, if the argument is null you should throw ArgumentNullException, BUT your code will throw NullReferenceException. Obviously, the exceptions vary and ArgumentNullException provides the most accurate information. It is in our interests that the ExceptionHandler receives the most accurate information.

mekk1t replied:
Interesting point. I'll give it some thought

mekk1t replied:
Fixed, added null checks.
Great idea, Mikhail.

MihailKulikov commented on DigitalOffice/Services/FileService/Models/FileRequest.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class FileRequest
+    {
+        public Guid Id { get; set; }
```
As far as I know, the request should not contain a unique file identifier. We issue it ourselves when writing to the database.

mekk1t replied:
As I mentioned above, we do not have Response model (at least I havent bump into any of them yet) so this FileRequest (I should have named it just File *sigh*) is a request and a response.
When we get the list of files, we get their data and their Ids so that we could click them and, for example, download by Id.

mekk1t replied:
Fixed Request model, removed Id property and created a separate Dto/Response model called "File" in "Models/" folder.

MihailKulikov commented on DigitalOffice/Services/FileService/Models/FileRequest.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Rename directory to RestRequests, because in Models directory we store DTO models

mekk1t replied:
Renamed

MihailKulikov commented on DigitalOffice/Services/FileService/Validators/NewFileValidator.cs:
```diff
@@ -0,0 +1,27 @@
+                .WithMessage("File name is too long")
+                .Matches("^[A-Z 0-9][A-Z a-z 0-9]+$");
+            RuleFor(file => file.Content)
+                .Must(f => IsBase64Coded(f));
```
Replace with method group.

MihailKulikov replied:
What happens if the Content is null?

mekk1t replied:
@MihailKulikov excuse me? Please, clarify what do you mean? What method group?
On the ContentIsNull -- good point. I should add null checks.

MihailKulikov replied:
You can write RuleFor(file => file.Content).Must(IsBase64Coded). It is little style fix, but it simplifies construction.

mekk1t replied:
Will check, thanks

mekk1t replied:
It's working and tests are passing. Thank you, Mikhail.

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/Commands/AddNewFileCommandTests.cs:
```diff
@@ -0,0 +1,109 @@
+                .Setup(x => x.AddNewFile(It.IsAny<DbFile>()))
+                .Returns(fileId);
+
+            Assert.AreEqual(fileId, command.Execute(fileRequest));
```
It would be a good idea to also check that when the command.Execute is called, the corresponding methods of the mocks are called.

mekk1t replied:
Could please advise me how to do this?
I cannot find it in your Pull Request (GetUser)

MihailKulikov replied:
Well, I have such check in my pull request. Example with mapper:
`mapperMock.Verify(mapper=> mapper.Map(fileReqeust), Times.Once);`
Similarly with other mocks.

mekk1t replied:
Huh. Weird, how'd I miss this. Thanks for info, will fix

mekk1t replied:
Fixed

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/Commands/AddNewFileCommandTests.cs:
```diff
@@ -0,0 +1,109 @@
+                 .Setup(x => x.Validate(It.IsAny<IValidationContext>()).IsValid)
+                 .Returns(false);
+
+            Assert.Throws<ValidationException>(() =>
```
Here also will be a good idea check that method of repositoryMock is not called.

mekk1t replied:
Added checks

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/Mappers/FileMapperTests.cs:
```diff
@@ -0,0 +1,91 @@
+            fileRequest = null;
+            dbFile = null;
+
+            Assert.Throws<NullReferenceException>(() => dbToRequestMapper.Map(dbFile));
```
It is bad practice to test two different methods in the same test.

mekk1t replied:
Fixed

mekk1t committed:
Models -> RestRequests

Renamed "Models" folder to "RestRequests", changed directives names.

mekk1t committed:
Added empty lines

mekk1t committed:
SerializerAssert class

Created a class that compares two JSON objects

mekk1t committed:
Update AddNewFileRepositoryTests.cs

Added a new test that fails when we add an existing item into the database.

mekk1t committed:
Updated Mapper Tests

Split mapping tests into two: DbToRequest, RequestToDb;
Used SerializerAssert when comparing mapping results.

mekk1t committed:
Update NewFileValidator.cs

Fixed Indentation
Added Null-check for the File Content.

mekk1t committed:
Merge branch 'develop' into feature/FileService_AddNewFile

Inzilbeth commented on DigitalOffice/Tests/FileServiceTests/Repositories/AddNewFileRepositoryTests.cs:
```diff
@@ -0,0 +1,60 @@
+        }
+
+        [Test]
+        public void SuccessfulAddNewFileToDatabaseTest()
+        {
+            Assert.AreEqual(newFile.Id, repository.AddNewFile(newFile));
+        }
```
This test only lets you know that the return Id is correct, but you don't check if the file actually appeared in the database

mekk1t replied:
Well, i merged Develop and cannot find how to check otherwise. Suggestions? 
All tests at the moment just check the returned values.
@Inzilbeth

mekk1t replied:
Ill try to find the item in the database by its id, if result is not null then we're cool

mekk1t replied:
Huh! My bad! I indeed added such check, but forgot to add it in Successful Test. Fixed!

mekk1t committed:
FIXES

1. Fixed request and dto mapping
2. REnamed directives adding LT.DigitalOffice.
3. Fixed tests, adding mock checks
4. Renamed variable names
4. Added null check and ArgumentNullException throwing when necessary.

Inzilbeth commented on DigitalOffice/Tests/FileServiceTests/Validators/AddNewFileValidatorTests.cs:
```diff
@@ -0,0 +1,50 @@
+        [Test]
+        public void SuccessfulFileValidationTest()
+        {
+            var fileValidationResult = validator.TestValidate(fileRequest);
+            fileValidationResult.ShouldNotHaveAnyValidationErrors();
```
wouldn't it be easier here and further in the validator tests to make it like:
`validator.ShouldNotHaveValidationErrorFor(request)`

mekk1t replied:
good idea

mekk1t replied:
@Inzilbeth but I cant't figure out how to pass LINQ as the first argument. 
So i used result variable and test it for no errors at all. So i leave it like that for the time being.

link from Slack: https://docs.fluentvalidation.net/en/latest/testing.html

KostinP commented on DigitalOffice/Services/FileService/Models/File.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace LT.DigitalOffice.FileService.Models
+{
+    public class File
```
What is it?

mekk1t replied:
Dto, as found in UserService from develop.

KostinP replied:
Hm... I thought we adding a _Response_ ending

KostinP commented:
Hm... We stopped adding a _Test_ ending for test methods...

KostinP commented on DigitalOffice/Tests/FileServiceTests/Commands/AddNewFileCommandTests.cs:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        [Test]
+        public void SuccessfulAddNewFileTest()
```
Hm... We stopped adding a _Test_ ending for test methods

djino97 replied:
At the end of the test name should be Test (convention).

mekk1t committed:
Fixed method declaration order

MihailKulikov commented on DigitalOffice/Services/FileService/Migrations/20200729053610_NewAddedOnProperty.cs:
```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace LT.DigitalOffice.FileService.Migrations
```
Migration should be in Database directory.

mekk1t replied:
Moved

MihailKulikov commented on DigitalOffice/Services/FileService/Migrations/20200729053610_NewAddedOnProperty.Designer.cs:
```diff
@@ -0,0 +1,50 @@
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace LT.DigitalOffice.FileService.Migrations
```
Migration should be in Database directory.

mekk1t replied:
Moved

MihailKulikov commented on DigitalOffice/Services/FileService/Mappers/FileMapper.cs:
```diff
@@ -0,0 +1,43 @@
+        {
+            if (dbFile == null)
+            {
+                throw new ArgumentNullException();
```
Why not add the argument name to the exception message? Use namof(dbFile)

mekk1t replied:
Fixed

MihailKulikov commented on DigitalOffice/Services/FileService/Mappers/FileMapper.cs:
```diff
@@ -0,0 +1,43 @@
+        {
+            if (file == null)
+            {
+                throw new ArgumentNullException();
```
Why not add the argument name to the exception message? Use namof(file)

mekk1t replied:
fixed

mekk1t committed:
Added validator call verification

mekk1t committed:
Repository Test: Fixed

MihailKulikov commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -16,26 +26,45 @@ public Startup(IConfiguration configuration)
+            ConfugureValidators(services);
+        }
+        
+        private void ConfugureCommands(IServiceCollection services)
```
Typo

mekk1t replied:
fixed

MihailKulikov commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -16,26 +26,45 @@ public Startup(IConfiguration configuration)
+            services.AddTransient<IAddNewFileCommand, AddNewFileCommand>();
         }
 
+        private void ConfugureRepositories(IServiceCollection services)
```
Typo

MihailKulikov commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -16,26 +26,45 @@ public Startup(IConfiguration configuration)
+            services.AddTransient<IFileRepository, FileRepository>();
+        }
+
+        private void ConfugureMappers(IServiceCollection services)
```
Typo

MihailKulikov commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -16,26 +26,45 @@ public Startup(IConfiguration configuration)
+            services.AddTransient<IMapper<FileCreateRequest, DbFile>, FileMapper>();
+        }
+
+        private void ConfugureValidators(IServiceCollection services)
```
Typo

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/Commands/AddNewFileCommandTests.cs:
```diff
@@ -0,0 +1,109 @@
+using LT.DigitalOffice.FileService.Repositories.Interfaces;
+using LT.DigitalOffice.FileService.Mappers.Interfaces;
+
+namespace FileServiceUnitTests.Commands
```
Fix namespace

mekk1t commented:
Updated everything, and everything works and tests fine.
The only problems are names of some files 
One comment is unresolved: it's Validator Test and "ShouldNotHaveAnyValidationErrors". I can't figure out what LINQ to pass, and the only one I've come up with is the properties and the request object as a Value. Screenshot attached.
![image](https://user-images.githubusercontent.com/61093487/89106489-eeea3080-d432-11ea-9da5-a30ea953873d.png)

mekk1t committed:
Update AddNewFileValidatorTests.cs

mekk1t commented:
UPD:
I managed to come up with an idea, but it's just a one-line economy. 
Any way, @Inzilbeth here: 
![image](https://user-images.githubusercontent.com/61093487/89106561-88194700-d433-11ea-8033-acb944be9f6a.png)

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/Mappers/FileMapperTests.cs:
```diff
@@ -0,0 +1,93 @@
+using LT.DigitalOffice.FileService.Mappers;
+using LT.DigitalOffice.FileService.Models;
+
+namespace FileServiceUnitTests.Mappers
```
Fix namespace

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/Repositories/AddNewFileRepositoryTests.cs:
```diff
@@ -0,0 +1,61 @@
+using NUnit.Framework;
+using System;
+
+namespace FileServiceUnitTests.Repositories
```
Fix namespace

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/UnitTestLibrary/SerializerAssert.cs:
```diff
@@ -0,0 +1,16 @@
+﻿using System.Text.Json;
+using NUnit.Framework;
+
+namespace FileServiceUnitTests.UnitTestLibrary
```
Fix namespace

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/Validators/AddNewFileValidatorTests.cs:
```diff
@@ -0,0 +1,49 @@
+using LT.DigitalOffice.FileService.Validators;
+using NUnit.Framework;
+
+namespace FileServiceUnitTests.Validators
```
Fix namespace

mekk1t committed:
Fixed namespaces

mekk1t committed:
Added nameof to Exceptions thrown

mekk1t commented:
@MihailKulikov fixed namespaces, typo and added nameof() to mappers

MihailKulikov commented on DigitalOffice/Services/FileService/Validators/NewFileValidator.cs:
```diff
@@ -0,0 +1,30 @@
+        {
+            RuleFor(file => file.Name)
+                .NotEmpty()
+                .WithMessage("File must have a name")
```
![изображение](https://user-images.githubusercontent.com/61084995/89108229-5b6b2c80-d43f-11ea-920e-3650c359d060.png)

mekk1t replied:
What does it mean? I wrote that it must have a name - so its not empty. Userfriendly isnt it

mekk1t replied:
Actually i didnt get what Pasha means, seems he wants to write "must not be empty", so I think "must have a name == must not be empty"

mekk1t committed:
Update AddNewFileRepositoryTests.cs

Added NotNull check in SuccessfulTest
